File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/E-Сортировка_вагонов/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>

using namespace std;


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    scanf("%d", &n);
    vector<int> nums (n, 0);
    for (int i = 0; i < n; ++i) cin >> nums[i];
    vector<string> moves;
    stack<int> deadlock;
    int rightNumsIdx = 0;
    int leftNumsIdx = 0;
    int target = 1;
    while (target <= n) {
        while (rightNumsIdx < n && nums[rightNumsIdx] != target) ++rightNumsIdx;
        
        if (rightNumsIdx < n) {
            moves.push_back("1 " + to_string(rightNumsIdx - leftNumsIdx + 1));
            for (int idx = leftNumsIdx; idx <= rightNumsIdx; ++idx) {
                deadlock.push(nums[idx]);
            }
            ++rightNumsIdx;
            leftNumsIdx = rightNumsIdx;
        } else {
            printf("0\n");
            return;
        }
        
        int deadlockToSecondStation = 0;
        while (!deadlock.empty() && deadlock.top() == target) {
            ++deadlockToSecondStation;
            deadlock.pop();
            ++target;
        }
        if (deadlockToSecondStation) {
            moves.push_back("2 " + to_string(deadlockToSecondStation));
        }
    }
    printf("%d\n", moves.size());
    for (string& move : moves) printf("%s\n", move.c_str());
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/E-Сортировка_вагонов/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    moves: list[str] = []
    deadlock: list[int] = []
    left: int = 0
    right: int = 0
    target: int = 1
    while target <= n:
        while right < n and nums[right] != target:
            right += 1
        if right < n:
            moves.append(f'1 {right - left + 1}')
            for idx in range(left, right + 1):
                deadlock.append(nums[idx])
            right += 1
            left = right
        else:
            sys.stdout.write('0\n')
            return
        
        deadlock_to_second_station: int = 0
        while deadlock and deadlock[-1] == target:
            deadlock_to_second_station += 1 
            target += 1 
            deadlock.pop()
        if deadlock_to_second_station:
            moves.append(f'2 {deadlock_to_second_station}')
    sys.stdout.write(f'{len(moves)}\n')
    sys.stdout.write('\n'.join(moves))
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution()